### ISSUE TITLE
Doesn't work with react-hot-loader (when inside redux-react-router)

### ISSUE THREAD

**Issue by qimingweng** [jsstyles/react-jss#18]
When a component is wrapped with @useSheet, it doesn't seem to work with `react-hot-loader` anymore. When I wrap a component with @useSheet and then again with a generic HOC, then it seems to work. I haven't investigated this very deeply but curious if others are experiencing similar things.

**Comment by qimingweng** [comment]
Maybe this will help: the error reads `TypeError: Cannot read property 'classes' of null`

(ie, the `this.props.sheet` object is `null`)

**Comment by gaearon** [comment]
@qimingweng Can I ask you to create a minimal boilerplate project reproducing the issue?

**Comment by qimingweng** [comment]
Sure, I'll try to do so as soon as I can

**Comment by gaearon** [comment]
Thanks. You can use https://github.com/gaearon/react-hot-boilerplate as template. In fact it would be awesome if you could make a PR adding `example` folder that just displays a single JSS component, based on https://github.com/gaearon/react-hot-boilerplate.

**Comment by qimingweng** [comment]
@gaearon upon further investigation this only seems to occur when the component is inside a `redux-react-router` instance.

**Comment by gaearon** [comment]
>upon further investigation this only seems to occur when the component is inside a redux-react-router instance.

Oh, thanks for clarifying. I remember @alexkuz telling me about this. 

@alexkuz Have you found why this was happening? (Or was it even you?)

**Comment by qimingweng** [comment]
@gaearon ah I see, would the error case still be helpful?

**Comment by gaearon** [comment]
@qimingweng Yes (although I'd ask you to raise issue in `redux-react-router` instead with it.) Thanks!

**Comment by alexkuz** [comment]
@qimingweng @gaearon I didn't trace the problem to its root, but I've managed to make it work. For some reason, during hot reload router is calling Route.component prop again, so if there is a component factory call, the component object will change. This breaks react-jss reference counter (componentWillUnmount() is called on the new component instead of previous) and things get ugly.
My workaround was to take component factory call out of Route prop, so this:
```
  React.render(
    <Provider store={store}>
    {() =>
      <Router history={history}>
        <Route component={reduxRouteComponent(store)} name='redux'>
          {routes()}
        </Route>
      </Router>
    }
    </Provider>,
    document.body
  );
```

becomes this:
```
const reduxRoute = reduxRouteComponent(store);

  React.render(
    <Provider store={store}>
    {() =>
      <Router history={history}>
        <Route component={reduxRoute} name='redux'>
          {routes()}
        </Route>
      </Router>
    }
    </Provider>,
    document.body
  );
```

**Comment by gaearon** [comment]
I’m closing as very old. Both `redux-router` and `react-hot-loader` have significantly changed since then.

**Comment by umidbekkarimov** [comment]
Faced this problem with `react-hot-loader@3.0.0-beta.5` with `react-jss@4.1.0`, 

Here how it goes:
1. Component Get Updated
2. [cWU](https://github.com/cssinjs/react-jss/blob/master/src/index.js#L64)
3. `refs` equals to `0` in [deref](https://github.com/cssinjs/react-jss/blob/master/src/index.js#L41), so it become `-1`
4. [cWM](https://github.com/cssinjs/react-jss/blob/master/src/index.js#L50)
5. `refs` equals to `-1` in [ref](https://github.com/cssinjs/react-jss/blob/master/src/index.js#L35) so `ref` doesn't attach sheets

**Comment by kof** [comment]
@umidbekkarimov  Can you create a pr with a test?

**Comment by umidbekkarimov** [comment]
@kof I can do something like this [react-redux/connect](https://github.com/reactjs/react-redux/blob/master/test/components/connect.spec.js#L1145)

**Comment by umidbekkarimov** [comment]
Finally, I caught this bug in testings,
I wanted to remove `[react-router] You cannot change <Router history>; it will be ignored` errors and used workaround from https://github.com/ReactTraining/react-router/issues/2704#issuecomment-211352123 
After this change every hot reload was starting from unmounting components, thats why refs get messed up. 

Fixed in #50
